File: backend-python-legacy/app/routers/weather.py
"""Rotas de clima (Open-Meteo): geocoding, config e dados."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException, Query

from app.providers.weather import (
    VALID_CURRENT,
    VALID_DAILY,
    VALID_HOURLY,
    VALID_PRECIPITATION_UNITS,
    VALID_TEMPERATURE_UNITS,
    VALID_WIND_UNITS,
    fetch_weather_data,
    mock_weather_payload,
    search_cities,
)
from app.schemas import WeatherConfig, WeatherGeocodingResponse, WeatherPatch, WeatherPayload
from app.store import load, update

router = APIRouter(prefix="/api/weather", tags=["weather"])


def _weather_config_public() -> WeatherConfig:
    cfg = load()
    raw = cfg.get("weather") or {}
    # Normaliza via store (já faz) e valida via Pydantic
    return WeatherConfig.model_validate(raw)


@router.get(
    "/config",
    response_model=WeatherConfig,
    summary="Configuração de clima",
)
def get_weather_config() -> WeatherConfig:
    return _weather_config_public()


@router.patch(
    "/config",
    response_model=WeatherConfig,
    summary="Atualiza configuração de clima",
)
def patch_weather_config(body: WeatherPatch) -> WeatherConfig:
    # Validações
    if body.temperature_unit is not None and body.temperature_unit not in VALID_TEMPERATURE_UNITS:
        raise HTTPException(400, f"temperature_unit inválido: {body.temperature_unit}")
    if body.wind_speed_unit is not None and body.wind_speed_unit not in VALID_WIND_UNITS:
        raise HTTPException(400, f"wind_speed_unit inválido: {body.wind_speed_unit}")
    if body.precipitation_unit is not None and body.precipitation_unit not in VALID_PRECIPITATION_UNITS:
        raise HTTPException(400, f"precipitation_unit inválido: {body.precipitation_unit}")
    if body.current is not None:
        invalid = [x for x in body.current if x not in VALID_CURRENT]
        if invalid:
            raise HTTPException(400, f"current inválido: {', '.join(invalid)}")
    if body.hourly is not None:
        invalid = [x for x in body.hourly if x not in VALID_HOURLY]
        if invalid:
            raise HTTPException(400, f"hourly inválido: {', '.join(invalid)}")
    if body.daily is not None:
        invalid = [x for x in body.daily if x not in VALID_DAILY]
        if invalid:
            raise HTTPException(400, f"daily inválido: {', '.join(invalid)}")

    def mut(cfg: dict[str, Any]) -> None:
        w = cfg.setdefault("weather", {})
        loc = w.setdefault("location", {})
        units = w.setdefault("units", {})
        disp = w.setdefault("display", {})
        fields = disp.setdefault("fields", {})

        if body.enabled is not None:
            w["enabled"] = body.enabled
            w["hidden"] = not body.enabled
        elif body.hidden is not None:
            w["hidden"] = body.hidden
            w["enabled"] = not body.hidden
        if body.name is not None:
            loc["name"] = body.name.strip()
        if body.latitude is not None:
            loc["latitude"] = body.latitude
        if body.longitude is not None:
            loc["longitude"] = body.longitude
        if body.country is not None:
            loc["country"] = body.country.strip()
        if body.country_code is not None:
            loc["country_code"] = body.country_code.strip()
        if body.timezone is not None:
            loc["timezone"] = body.timezone.strip() or "auto"
            w["timezone"] = body.timezone.strip() or "auto"
        if body.elevation is not None:
            loc["elevation"] = body.elevation
        if body.temperature_unit is not None:
            units["temperature_unit"] = body.temperature_unit
        if body.wind_speed_unit is not None:
            units["wind_speed_unit"] = body.wind_speed_unit
        if body.precipitation_unit is not None:
            units["precipitation_unit"] = body.precipitation_unit
        if body.forecast_days is not None:
            w["forecast_days"] = body.forecast_days
        if body.past_days is not None:
            w["past_days"] = body.past_days
        if body.current is not None:
            w["current"] = body.current
        if body.hourly is not None:
            w["hourly"] = body.hourly
        if body.daily is not None:
            w["daily"] = body.daily
        if body.display_show_current is not None:
            disp["show_current"] = body.display_show_current
        if body.display_show_hourly is not None:
            disp["show_hourly"] = body.display_show_hourly
        if body.display_show_daily is not None:
            disp["show_daily"] = body.display_show_daily
        if body.display_hourly_count is not None:
            disp["hourly_count"] = body.display_hourly_count
        if body.display_daily_count is not None:
            disp["display_daily_count"] = body.display_daily_count
            disp["daily_count"] = body.display_daily_count
        if body.display_fields is not None:
            for k, v in body.display_fields.items():
                if k in fields:
                    fields[k] = bool(v)

    update(mut)
    return _weather_config_public()


@router.get(
    "/geocoding",
    response_model=WeatherGeocodingResponse,
    summary="Busca cidades (proxy Geocoding Open-Meteo)",
)
def geocoding_search(
    q: str = Query(..., min_length=2, description="Nome da cidade"),
    count: int = Query(5, ge=1, le=10),
    language: str = Query("pt"),
) -> WeatherGeocodingResponse:
    try:
        data = search_cities(q, count=count, language=language)
    except RuntimeError as exc:
        raise HTTPException(502, str(exc)) from exc
    return WeatherGeocodingResponse.model_validate(data)


@router.get(
    "",
    response_model=WeatherPayload,
    summary="Dados meteorológicos atuais (força fetch)",
)
def get_weather() -> WeatherPayload:
    cfg = load()
    wcfg = cfg.get("weather") or {}
    if cfg.get("mock") and wcfg.get("enabled"):
        return WeatherPayload.model_validate(mock_weather_payload())
    data = fetch_weather_data(wcfg)
    return WeatherPayload.model_validate(data)


@router.post(
    "/location",
    response_model=WeatherConfig,
    summary="Define localização por cidade (atalho)",
)
def set_location(body: dict[str, Any]) -> WeatherConfig:
    name = str(body.get("name") or "").strip()
    lat = body.get("latitude")
    lon = body.get("longitude")
    if lat is None or lon is None:
        raise HTTPException(400, "latitude e longitude são obrigatórios")
    try:
        lat_f = float(lat)
        lon_f = float(lon)
    except (TypeError, ValueError):
        raise HTTPException(400, "latitude/longitude inválidos")
    if not (-90 <= lat_f <= 90 and -180 <= lon_f <= 180):
        raise HTTPException(400, "latitude/longitude fora do intervalo")

    def mut(cfg: dict[str, Any]) -> None:
        w = cfg.setdefault("weather", {})
        loc = w.setdefault("location", {})
        loc["name"] = name
        loc["latitude"] = lat_f
        loc["longitude"] = lon_f
        if body.get("country") is not None:
            loc["country"] = str(body["country"]).strip()
        if body.get("country_code") is not None:
            loc["country_code"] = str(body["country_code"]).strip()
        if body.get("timezone") is not None:
            tz = str(body["timezone"]).strip() or "auto"
            loc["timezone"] = tz
            w["timezone"] = tz
        if body.get("elevation") is not None:
            try:
                loc["elevation"] = float(body["elevation"])
            except (TypeError, ValueError):
                pass
        # Ao definir localização, liga o clima no painel (busca + card).
        w["enabled"] = True
        w["hidden"] = False

    update(mut)
    return _weather_config_public()

File: backend/src/routers/weather.ts
import type { FastifyInstance } from "fastify";
import {
  VALID_CURRENT,
  VALID_DAILY,
  VALID_HOURLY,
  VALID_TEMPERATURE_UNITS,
  VALID_WIND_UNITS,
  VALID_PRECIPITATION_UNITS,
  fetchWeatherData,
  mockWeatherPayload,
  searchCities,
} from "../providers/weather.js";
import { load, updateSync as update } from "../store.js";
import { imageToRaw } from "./wallpapers/rgb565.js";
import { downloadImage } from "./wallpapers/ssrfGuard.js";

// Mesmo emoji de WMO_EMOJI (frontend/src/format.ts), só que pelo nome do
// arquivo do Twemoji (codepoint sem o FE0F) — a ESP32 não desenha emoji.
const WMO_TWEMOJI: Record<number, string> = {
  0: "2600", 1: "1f324", 2: "26c5", 3: "2601",
  45: "1f32b", 48: "1f32b",
  51: "1f326", 53: "1f326", 55: "1f327", 56: "1f327", 57: "1f327",
  61: "1f327", 63: "1f327", 65: "1f327", 66: "1f327", 67: "1f327",
  71: "1f328", 73: "1f328", 75: "2744", 77: "2744",
  80: "1f326", 81: "1f326", 82: "26c8", 85: "1f328", 86: "2744",
  95: "26c8", 96: "26c8", 99: "26c8",
};
const TWEMOJI_FALLBACK = "1f321"; // 🌡️
const TWEMOJI_BASE = "https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/72x72/";
const ICON_MIN = 16;
const ICON_MAX = 72;
const ICON_DEFAULT = 48;
// Cor-chave de transparência: a mesma kBakedCard do firmware (ui/customtheme.cpp).
const DEFAULT_KEY = 0x1904;

export function twemojiForWmo(code: number | null): string {
  if (code === null) return TWEMOJI_FALLBACK;
  return WMO_TWEMOJI[code] ?? TWEMOJI_FALLBACK;
}

const iconCache = new Map<string, Buffer>();

export async function createWeatherRoutes(app: FastifyInstance): Promise<void> {
  // Ícone da condição atual em RAW RGB565 little-endian — mesma estratégia da
  // capa do Spotify (/api/spotify/cover): o coletor baixa e converte, a placa
  // só faz pushImage. A placa manda o `code` que já recebeu no /usage, então a
  // rota não depende do ciclo do hub e o cache fica por código.
  app.get("/api/weather/icon", { schema: { tags: ["Clima"] } }, async (request, reply) => {
    const q = (request.query ?? {}) as Record<string, string | undefined>;
    const rawCode = q.code === undefined || q.code === "" ? NaN : Number(q.code);
    const code = Number.isInteger(rawCode) ? rawCode : null;
    const rawSize = Number(q.size ?? ICON_DEFAULT);
    const size = Number.isFinite(rawSize) ? Math.min(ICON_MAX, Math.max(ICON_MIN, Math.round(rawSize))) : ICON_DEFAULT;
    const rawKey = q.key !== undefined ? parseInt(q.key, 16) : DEFAULT_KEY;
    const key = Number.isInteger(rawKey) && rawKey >= 0 && rawKey <= 0xffff ? rawKey : DEFAULT_KEY;
    const file = twemojiForWmo(code);
    const cacheKey = `${file}:${size}:${key}`;
    try {
      let raw = iconCache.get(cacheKey);
      if (!raw) {
        const png = await downloadImage(`${TWEMOJI_BASE}${file}.png`);
        raw = await imageToRaw(png, size, size, { transparentKey: key });
        iconCache.set(cacheKey, raw);
      }
      return reply
        .header("Content-Type", "application/octet-stream")
        .header("X-Vigia-Weather-Code", code === null ? "" : String(code))
        .header("X-Vigia-Icon-Size", String(size))
        .header("Cache-Control", "no-store")
        .send(raw);
    } catch (e) {
      return reply.code(502).send({ ok: false, error: String((e as Error).message ?? e) });
    }
  });

  app.get("/api/weather/config", { schema: { tags: ["Clima"] } }, async () => {
    const cfg = load() as Record<string, unknown>;
    const raw = (cfg.weather ?? {}) as Record<string, unknown>;
    return raw;
  });

  app.patch("/api/weather/config", { schema: { tags: ["Clima"] } }, async (request, reply) => {
    const body = request.body as Record<string, unknown> | null;
    if (!body) return reply.code(400).send({ ok: false, error: "corpo vazio" });
    if (body.temperature_unit !== undefined && body.temperature_unit !== null && !VALID_TEMPERATURE_UNITS.has(String(body.temperature_unit))) {
      return reply.code(400).send({ ok: false, error: `temperature_unit inválido: ${body.temperature_unit}` });
    }
    if (body.wind_speed_unit !== undefined && body.wind_speed_unit !== null && !VALID_WIND_UNITS.has(String(body.wind_speed_unit))) {
      return reply.code(400).send({ ok: false, error: `wind_speed_unit inválido: ${body.wind_speed_unit}` });
    }
    if (body.precipitation_unit !== undefined && body.precipitation_unit !== null && !VALID_PRECIPITATION_UNITS.has(String(body.precipitation_unit))) {
      return reply.code(400).send({ ok: false, error: `precipitation_unit inválido: ${body.precipitation_unit}` });
    }
    if (body.current !== undefined && body.current !== null) {
      const invalid = (body.current as string[]).filter((x) => !VALID_CURRENT.has(x));
      if (invalid.length) return reply.code(400).send({ ok: false, error: `current inválido: ${invalid.join(", ")}` });
    }
    if (body.hourly !== undefined && body.hourly !== null) {
      const invalid = (body.hourly as string[]).filter((x) => !VALID_HOURLY.has(x));
      if (invalid.length) return reply.code(400).send({ ok: false, error: `hourly inválido: ${invalid.join(", ")}` });
    }
    if (body.daily !== undefined && body.daily !== null) {
      const invalid = (body.daily as string[]).filter((x) => !VALID_DAILY.has(x));
      if (invalid.length) return reply.code(400).send({ ok: false, error: `daily inválido: ${invalid.join(", ")}` });
    }

    update((cfg: Record<string, unknown>) => {
      const w = (cfg.weather ?? {}) as Record<string, unknown>;
      if (!cfg.weather) cfg.weather = w;
      const loc = (w.location ?? {}) as Record<string, unknown>;
      if (!w.location) w.location = loc;
      const units = (w.units ?? {}) as Record<string, unknown>;
      if (!w.units) w.units = units;
      const disp = (w.display ?? {}) as Record<string, unknown>;
      if (!w.display) w.display = disp;
      const fields = (disp.fields ?? {}) as Record<string, unknown>;
      if (!disp.fields) disp.fields = fields;

      if (body.enabled !== undefined && body.enabled !== null) { w.enabled = Boolean(body.enabled); w.hidden = !Boolean(body.enabled); }
      else if (body.hidden !== undefined && body.hidden !== null) { w.hidden = Boolean(body.hidden); w.enabled = !Boolean(body.hidden); }
      if (body.name !== undefined && body.name !== null) loc.name = String(body.name).trim();
      if (body.latitude !== undefined && body.latitude !== null) loc.latitude = Number(body.latitude);
      if (body.longitude !== undefined && body.longitude !== null) loc.longitude = Number(body.longitude);
      if (body.country !== undefined && body.country !== null) loc.country = String(body.country).trim();
      if (body.country_code !== undefined && body.country_code !== null) loc.country_code = String(body.country_code).trim();
      if (body.timezone !== undefined && body.timezone !== null) { const tz = String(body.timezone).trim() || "auto"; loc.timezone = tz; w.timezone = tz; }
      if (body.elevation !== undefined && body.elevation !== null) loc.elevation = Number(body.elevation);
      if (body.temperature_unit !== undefined && body.temperature_unit !== null) units.temperature_unit = String(body.temperature_unit);
      if (body.wind_speed_unit !== undefined && body.wind_speed_unit !== null) units.wind_speed_unit = String(body.wind_speed_unit);
      if (body.precipitation_unit !== undefined && body.precipitation_unit !== null) units.precipitation_unit = String(body.precipitation_unit);
      if (body.forecast_days !== undefined && body.forecast_days !== null) w.forecast_days = Number(body.forecast_days);
      if (body.past_days !== undefined && body.past_days !== null) w.past_days = Number(body.past_days);
      if (body.current !== undefined && body.current !== null) w.current = body.current;
      if (body.hourly !== undefined && body.hourly !== null) w.hourly = body.hourly;
      if (body.daily !== undefined && body.daily !== null) w.daily = body.daily;
      if (body.display_show_current !== undefined && body.display_show_current !== null) disp.show_current = Boolean(body.display_show_current);
      if (body.display_show_hourly !== undefined && body.display_show_hourly !== null) disp.show_hourly = Boolean(body.display_show_hourly);
      if (body.display_show_daily !== undefined && body.display_show_daily !== null) disp.show_daily = Boolean(body.display_show_daily);
      if (body.display_hourly_count !== undefined && body.display_hourly_count !== null) disp.hourly_count = Number(body.display_hourly_count);
      if (body.display_daily_count !== undefined && body.display_daily_count !== null) { disp.display_daily_count = Number(body.display_daily_count); disp.daily_count = Number(body.display_daily_count); }
      if (body.display_fields !== undefined && body.display_fields !== null) {
        for (const [k, v] of Object.entries(body.display_fields as Record<string, unknown>)) {
          if (k in fields) fields[k] = Boolean(v);
        }
      }
    });
    const cfg = load() as Record<string, unknown>;
    return (cfg.weather ?? {}) as Record<string, unknown>;
  });

  app.get("/api/weather/geocoding", { schema: { tags: ["Clima"] } }, async (request, reply) => {
    const query = (request.query ?? {}) as Record<string, string>;
    const q = String(query.q ?? "");
    const count = Number(query.count ?? 5);
    const language = String(query.language ?? "pt");
    if (!q || q.trim().length < 2) return reply.code(400).send({ ok: false, error: "q deve ter pelo menos 2 caracteres" });
    try {
      const data = await searchCities(q, count, language);
      return data;
    } catch (exc) {
      return reply.code(502).send({ ok: false, error: String(exc) });
    }
  });

  app.get("/api/weather", { schema: { tags: ["Clima"] } }, async () => {
    const cfg = load() as Record<string, unknown>;
    const wcfg = (cfg.weather ?? {}) as Record<string, unknown>;
    if (cfg.mock && wcfg.enabled) return mockWeatherPayload();
    const data = await fetchWeatherData(wcfg);
    return data;
  });

  app.post("/api/weather/location", { schema: { tags: ["Clima"] } }, async (request, reply) => {
    const body = request.body as Record<string, unknown> | null;
    if (!body) return reply.code(400).send({ ok: false, error: "corpo vazio" });
    const name = String(body.name ?? "").trim();
    const lat = body.latitude;
    const lon = body.longitude;
    if (lat === null || lat === undefined || lon === null || lon === undefined) return reply.code(400).send({ ok: false, error: "latitude e longitude são obrigatórios" });
    let latF: number, lonF: number;
    try { latF = Number(lat); lonF = Number(lon); if (Number.isNaN(latF) || Number.isNaN(lonF)) throw new Error(); } catch { return reply.code(400).send({ ok: false, error: "latitude/longitude inválidos" }); }
    if (!( -90 <= latF && latF <= 90 && -180 <= lonF && lonF <= 180)) return reply.code(400).send({ ok: false, error: "latitude/longitude fora do intervalo" });
    update((cfg: Record<string, unknown>) => {
      const w = (cfg.weather ?? {}) as Record<string, unknown>;
      if (!cfg.weather) cfg.weather = w;
      const loc = (w.location ?? {}) as Record<string, unknown>;
      if (!w.location) w.location = loc;
      loc.name = name;
      loc.latitude = latF;
      loc.longitude = lonF;
      if (body.country !== undefined && body.country !== null) loc.country = String(body.country).trim();
      if (body.country_code !== undefined && body.country_code !== null) loc.country_code = String(body.country_code).trim();
      if (body.timezone !== undefined && body.timezone !== null) { const tz = String(body.timezone).trim() || "auto"; loc.timezone = tz; w.timezone = tz; }
      if (body.elevation !== undefined && body.elevation !== null) { try { loc.elevation = Number(body.elevation); } catch {} }
      w.enabled = true;
      w.hidden = false;
    });
    const cfg = load() as Record<string, unknown>;
    return (cfg.weather ?? {}) as Record<string, unknown>;
  });
}
